hathor/event/websocket/protocol.py

```diff
@@ -0,0 +1,186 @@
+
+logger = get_logger()
+
+
+class EventWebsocketProtocol(WebSocketServerProtocol):
+    """ Websocket protocol, basically forwards some events to the Websocket factory.
+    """
+
+    factory: 'EventWebsocketFactory'
+    client_peer: Optional[str] = None
+
+    _last_sent_event_id: Optional[int] = None
+    _ack_event_id: Optional[int] = None
+    _window_size: int = 0
+    _stream_is_active: bool = False
+
+    def __init__(self):
+        super().__init__()
+        self.log = logger.new()
+
+    def can_receive_event(self, event_id: int) -> bool:
+        """Returns whether this client is available to receive an event."""
+        number_of_pending_events = 0
+
+        if self._last_sent_event_id is not None:
+            ack_offset = -1 if self._ack_event_id is None else self._ack_event_id
+            number_of_pending_events = self._last_sent_event_id - ack_offset
+
+        return (
+            self._stream_is_active
+            and event_id == self.next_expected_event_id()
+            and number_of_pending_events < self._window_size
+        )
+
+    def next_expected_event_id(self) -> int:
+        """Returns the ID of the next event the client expects."""
+        return 0 if self._last_sent_event_id is None else self._last_sent_event_id + 1
+
+    def onConnect(self, request: ConnectionRequest) -> None:
+        self.client_peer = request.peer
+        self.log = self.log.new(client_peer=self.client_peer)
+        self.log.info('connection opened to the event websocket, starting handshake...')
+
+    def onOpen(self) -> None:
+        self.log.info('connection established to the event websocket')
+        self.factory.register(self)
+
+    def onClose(self, wasClean: bool, code: int, reason: str) -> None:
+        self.log.info('connection closed to the event websocket', reason=reason)
+        self.factory.unregister(self)
+
+    def onMessage(self, payload: bytes, isBinary: bool) -> None:
+        self.log.debug('message', payload=payload.hex() if isBinary else payload.decode('utf8'))
+
+        try:
+            request = RequestWrapper.parse_raw_request(payload)
+            self._handle_request(request)
+        except ValidationError as error:
+            self.send_invalid_request_response(InvalidRequestType.VALIDATION_ERROR, payload, str(error))
+        except InvalidRequestError as error:
+            self.send_invalid_request_response(error.type, payload)
+
+    def _handle_request(self, request: Request) -> None:
+        # This could be a pattern match in Python 3.10
+        request_type = type(request)
+        handlers: Dict[Type, Callable] = {
+            StartStreamRequest: self._handle_start_stream_request,
+            AckRequest: self._handle_ack_request,
+            StopStreamRequest: lambda _: self._handle_stop_stream_request()
+        }
+        handle_fn = handlers.get(request_type)
+
+        assert handle_fn is not None, f'cannot handle request of unknown type "{request_type}"'
```

Comment:
Should we raise a InvalidRequestError here? I feel an assert would be considered an unhandled exception and might open incidents for the on call.

Reply:
> Or would pydantic consider the request invalid already?

Yep, that's correct. Pydantic's parsing happens on the `onMessage()` method:

```python
request = RequestWrapper.parse_raw_request(payload)
```

If a request is provided with a nonexistent type, a `ValidationError` will be raised and handled, sending a `InvalidRequestResponse` to the client. This is defined by the `Request` type alias

```python
Request = Union[StartStreamRequest, AckRequest, StopStreamRequest]
```

and the respective `type` property type hint of each request type, for example: `type: Literal['START_STREAM']`.

If the code reaches this point and raises that assertion, it means that some dev added a new request type (making it parsable, and therefore that request would be valid) but forgot to implement the handler for that type.

In Python 3.10, that whole method would be changed to:

```python
def _handle_request(self, request: Request) -> None:
    match request:
        case StartStreamRequest():
            self._handle_start_stream_request(request)
        case AckRequest():
            self._handle_ack_request(request)
        case StopStreamRequest():
            self._handle_stop_stream_request()
```

Reply:
Got it! Thanks for the explanation.